README.md
# node-dota-api
A tool to fetch Dota stats for players, heroes, matches, and more!

Based off of [OpenDota's new API](https://docs.opendota.com/) & Dota 2's Web API

## What's New? ##

## v4.0.1

- Minor bugfix for mistyped conditoinal check (thanks [Quazia](https://github.com/ammuench/node-dota-api/pull/4))

### v4.0.0

- Update all callbacks to now be promises (thanks [Quazia](https://github.com/ammuench/node-dota-api/pull/3))

### v3.0.1

- Add new prime account check status

[See previous changelogs here](https://github.com/ammuench/node-dota-api/blob/master/CHANGELOG.md)


## Installation ##
------

You can install through NPM by running
```bash
npm install node-dota-api
```

or by cloning from git into your project
```bash
git clone https://github.com/ammuench/node-dota-api.git
```

## API ##
------
##### nodeDotaApi.playerStats(*playerID*)
Takes a PlayerID value as a string or integer. Returns a promise that passes back a playerInfo JSON object with the following format:
```json
{
    "status": "",
    "daysSinceLastMatch": "",
    "soloMMR": "",
    "partyMMR": "",
    "estMMR": "",
    "name": "",
    "winLoss": {
        "wins": "",
        "losses": "",
        "winrate": "",
        "totalGames": ""
    },
    "profileURL": "",
    "profileImage": "",
    "isPrime" : BOOLEAN or "No apiKey provided"
};
```
###### Status Reports

If the user is valid, but their profile shows less than 20 games, the status will return as "**Partial Match History**".  This is likely due to match history not being shown, or an extremely new account.

If the user is valid, but their profile's last parsed game is over 14 days ago, the status will return as "**Outdated Match History**".

If the user has NO match history at all, the status will return as "**No Match History/Invalid Account**", as either are possible (invalid ID number, or no exposed data).

If there an invalid value passed in, status will return as "**Invalid**". 

If it is a server error, status will return as "**Error**".

Prime Check sees if a ranked game was played after timestamp "1493948227" (roughly around when the prime update went live).

## License ##
-----
(The MIT License)

Copyright (c) 2016 Alex Muench <ammuench@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

Users of this software will give the developer at least one ('1') high-five upon meeting him in person as a sign of gratitude.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Special Thanks ##

Shoutout to the guys at [OpenDota](http://opendota.com), they do great work, and they're the reason this API works.  Go [donate and help carry](https://www.opendota.com/carry) them to keep their systems running!

Thanks to [Arthur](https://github.com/quazia) for his contributions to the project

Many thanks to Carney for answer myriads of questions when I'm having issues

Much love to Elizabeth for wanting me to win at my code <3



CHANGELOG.md
# v4.0.1

- Minor bugfix for mistyped conditional check

## v4.0.0

- Update all callbacks to promises

## v3.0.1

- Add new prime account check status

## v2.0.1

- Quick fix to make the `daysSinceLastMatch` value more accurate after talking with the API dev

## v2.0.0

- Moving to becoming a wrapper for OpenDota's new API
- `mostPlayed` and `recentGame` values currently not accessible (will be re-added soon!)

-----

## v1.0.0

- Major stable release!
- Ported to use opendota instead of yasp
- Added in game length and game type
- Expanded status codes to be more descriptive
- Added 'daysSinceLastMatch' value
- Added total games played to 'winLoss' object
- Locked dependency versions for stability

-----

## v0.2.4

- Added 'profileImage' to playerStats object

## v0.2.3

- Quick hotfix for a 0.2.2 issue

## v0.2.2

- Added check for invalid user
- whenPlayed now returns a formatted date string
- kda added to recent match objects

## v0.2.1

- Added match history to playerStats object

## v0.2

- Stripped out all the express jargon from 0.1
- Flushed out full playerStats service

server.js
var express = require('express');
var fs = require('fs');
var request = require('request');
var cheerio = require('cheerio');
var app = express();

app.all('*', function(req, res, next){
  if (!req.get('Origin')) return next();
  // use "*" here to accept any origin
  res.set('Access-Control-Allow-Origin', '*');
  res.set('Access-Control-Allow-Methods', 'GET, POST, PUT, OPTIONS');
  res.set('Access-Control-Allow-Headers', 'X-Requested-With, Content-Type, Authentication, AdminAccess');
  if ('OPTIONS' == req.method) return res.send(200);
  next();
});



app.get('/playerinfo/:id', function(req, res){
	var profileId = req.params.id;

	var profileUrl = 'https://yasp.co/players/' + req.params.id;

	var userInfoJson = { soloMMR : "", partyMMR : "", estMMR: "", name: "", profileURL: profileUrl};

	request(profileUrl, function(error, response, html){
		if(!error){
			var $ = cheerio.load(html);

			var soloMMR, teamMMR, estMMR, name;


			$('#estimate').filter(function(){

				estMMR = $(this).children('span').children('abbr').text();

				userInfoJson.estMMR = estMMR;

			});

			$('abbr[title="Solo MMR"]').filter(function(){

				soloMMR = $(this).parent().next().text();

				userInfoJson.soloMMR = soloMMR;

			});

			$('abbr[title="Party MMR"]').filter(function(){

				partyMMR = $(this).parent().next().text();

				userInfoJson.partyMMR = partyMMR;

			});

			$('title').filter(function(){

				console.log($(this));
				name = $(this).text().split(' - YASP')[0];

				userInfoJson.name = name;

			});

			res.send(userInfoJson);
		}else{
			console.log('Error parsing userInfo URL');
		}
		
	});
});


var port = process.env.PORT || 8081;

app.listen(port);
console.log('Extra-Life API running on port ' + port);
exports = module.exports = app;

dotaPlayerApi.js
/**
 * node-dota-api
 * A node-based API to get Dota stats for players, heroes, matches, and more!
 * @version 0.2.4
 * @author Alex Muench
 */

var moment = require('moment');

//define requirements
var request = require('request');
var cheerio = require('cheerio');

module.exports = {

	/**
	 * [playerStats description]
	 * @param  {string|integer}	playerID							Player's Dota ID Number
	 * @property {object}		playerInfoJson						JSON object containing player stats and info
	 * @property {string}		playerInfoJson.status				Status of request call.  See status reports section in README.md
	 * @property {float}		playerInfoJson.daysSinceLastMatch	Number of days since last match was played && parsed
	 * @property {string}		playerInfoJson.soloMMR				Solo MMR Value
	 * @property {string}		playerInfoJson.partyMMR				Party MMR Value
	 * @property {string}		playerInfoJson.estMMR				OpenDota Estimated MMR Value
	 * @property {string}		playerInfoJson.name					User's current Profile Name
	 * @property {object}		playerInfoJson.winLoss				Object of Win-Loss values
	 * @property {string}		playerInfoJson.winLoss.wins			Number of wins (returned as string)
	 * @property {string}		playerInfoJson.winLoss.losses		Number of losses (returned as string)
	 * @property {string}		playerInfoJson.winLoss.winrate		Winrate percentage
	 * @property {string}		playerInfoJson.winLoss.totalGames	Number of total games played & parsed
	 * @property {array}		playerInfoJson.mostPlayed			Array of User's 5 most played heroes, with games played and winrate
	 * @property {array}		playerInfoJson.recentGames			Array of User's 20 most recent games, with matchID, hero, result, time played, and skill level
	 * @property {string}		playerInfoJson.profileUrl			URL to user's proflie on OpenDota.com
	 */
	playerStats: function (playerID, apiKey) {
		var apiBase = 'https://api.opendota.com/api/players/' + playerID;

		var playerInfoJson = {
			status: '',
			daysSinceLastMatch: '',
			soloMMR: '',
			partyMMR: '',
			estMMR: '',
			name: '',
			winLoss: {
				wins: '',
				losses: '',
				winrate: '',
				totalGames: ''
			},
			mostPlayed: [],
			recentGames: [],
			profileURL: 'https://opendota.com/players/' + playerID,
			profileImage: '',
			isPrime: ''
		};
		return new Promise(function (resolve, reject) {
			request(apiBase, function (err, res) {
				if (err) {
					playerInfoJson.status = 'Error';
					reject("error with initial request");
				} else {
					const data = JSON.parse(res.body);
					if (data.error) {
						playerInfoJson.status = 'Invalid';
						resolve(playerInfoJson);
					} else {
						playerInfoJson.status = 'Valid Account';
						playerInfoJson.soloMMR = data.solo_competitive_rank;
						playerInfoJson.partyMMR = data.competitive_rank;
						playerInfoJson.estMMR = data.mmr_estimate.estimate;
						playerInfoJson.name = data.profile.personaname;
						playerInfoJson.profileImage = data.profile.avatarfull;

						request(apiBase + '/wl', function (err, res) {
							if (err) {
								playerInfoJson.status = 'Error';
								reject("error with wl request");
							}
							const data = JSON.parse(res.body);
							if(data.lose == null || data.win == null){
								reject("error fetching win loss data");
							}
							playerInfoJson.winLoss.losses = data.lose;
							playerInfoJson.winLoss.wins = data.win;
							playerInfoJson.winLoss.totalGames = data.lose + data.win;
							playerInfoJson.winLoss.winrate = (data.win / (data.lose + data.win));

							request(apiBase + '/matches?limit=1', function (err, res) {
								if (err) {
									playerInfoJson.status = 'Error';
									reject("error with wl request");
								}
								const data = JSON.parse(res.body)[0];
								var now = moment(new Date());
								//temp fix while the data returned is off by one month
								if(data.start_time == null){
									reject("Error fetching start time of last match");
								}
								var lastGame = moment.unix(data.start_time)
								var duration = moment.duration(now.diff(lastGame)).asDays();
								playerInfoJson.daysSinceLastMatch = duration;
								if (duration > 14) {
									playerInfoJson.status = 'Outdated Match History'
								}

								if (!apiKey) {
									playerInfoJson.isPrime = 'No apiKey provided';
									resolve(playerInfoJson);
								} else {
									const steamwebapi = `https://api.steampowered.com/IDOTA2Match_570/GetMatchHistory/v1/?key=${apiKey}&matches_requested=30&account_id=${playerID}&game_mode=7&format=JSON`;
									let isPrime = false;
									request(steamwebapi, function (err, res) {
										if (err) {
											playerInfoJson.status = 'Error';
											reject("error with wl request");
										}
										const matches = JSON.parse(res.body).result.matches;
										if(matches == null){
											reject("error fetching matchs");
										}
										for (let match of matches) {
											if (match.lobby_type === 7) {
												isPrime = (match.start_time > 1493948227);
												break;
											}
										}
										playerInfoJson.isPrime = isPrime;
										resolve(playerInfoJson);
									});
								}

							})

						})
					}
				}
			});
		});
	}

};
